eva-tools/src/main/java/embl/ebi/variation/eva/vcfDump/VariantExporter.java

```diff
@@ -0,0 +1,350 @@
+                                            VariantSourceDBAdaptor sourceDBAdaptor, QueryOptions options) throws IOException {
+
+
+        List<String> studyIds = options.getAsStringList(VariantDBAdaptor.STUDIES);
+        List<VariantSource> sourcesList = sourceDBAdaptor.getAllSourcesByStudyIds(studyIds, options).getResult();
+
+        Map<String, VariantSource> sources = new TreeMap<>();
+
+        for (VariantSource variantSource : sourcesList) {
+            sources.put(variantSource.getStudyId(), variantSource);
+        }
+        
+        Map<String, VCFHeader> headers = getVcfHeaders(sources);
+//        header.addMetaDataLine(new VCFFilterHeaderLine("PASS", "Valid variant"));
+//        header.addMetaDataLine(new VCFFilterHeaderLine(".", "No FILTER info"));
+
+        String suffix = ".exported.vcf.gz";
+        List<String> files = new ArrayList<>();
+        Map<String, VariantContextWriter> writers = new TreeMap<>();
+
+        // setup writers
+        for (String studyId : studyIds) {
+            if (!headers.containsKey(studyId)) {
+                throw new IllegalArgumentException("aborting VCF export: missing header for study " + studyId);
+            }
+            VariantContextWriterBuilder builder = new VariantContextWriterBuilder();
+            File outFile = Paths.get(outputDir).resolve(studyId + suffix).toFile();
+            files.add(outFile.getPath());
+            VariantContextWriter writer = builder
+                    .setOutputFile(outFile)
+                    .setReferenceDictionary(headers.get(studyId).getSequenceDictionary())
+                    .unsetOption(Options.INDEX_ON_THE_FLY)
+                    .build();
+            writers.put(studyId, writer);
+            writer.writeHeader(headers.get(studyId));
+        }
+
+        // actual loop
+        int failedVariants = 0;
+        StringBuilder filesConcat = new StringBuilder();
+        for (String file : files) {
+            filesConcat.append(file).append(" ");
+        }
+        logger.info("Exporting to files: [ " + filesConcat.toString() + "]");
+
+        while (iterator.hasNext()) {
+            Variant variant = iterator.next();
+            try {
+                Map<String, VariantContext> variantContexts = convertBiodataVariantToVariantContext(variant, sources);
+                for (Map.Entry<String, VariantContext> variantContextEntry : variantContexts.entrySet()) {
+                    if (writers.containsKey(variantContextEntry.getKey())) {
+                        writers.get(variantContextEntry.getKey()).add(variantContextEntry.getValue());
+                    }
+                }
+            } catch (Exception e) {
+                logger.info("failed variant: ", e);
+                failedVariants++;
+            }
+        }
+
+        if (failedVariants > 0) {
+            logger.warn(failedVariants + " variants were not written due to errors");
+        }
+        VariantExporter.failedVariants += failedVariants;
```

Comment:
I like the idea of keeping a count of failed variants :smile: But if I understand the code well, the way it works right now, it will accumulate the failed variants from all runs, but the count in a single run can't be retrieved by other parts of the system; it will only be displayed in the log.

It could be interesting to turn it into an attribute which could be accessed via a getter method, so an external component could react accordingly. For instance, rejecting to make files available in case of errors, display the number of failed variants in a website or email, etc.

This would most probably also mean making the methods non-static, as we had already considered.